src/vm/runners/builtin_runner/signature.rs

```diff
@@ -0,0 +1,469 @@
+            cells_per_instance: 5,
+            n_input_cells: 2,
+            _total_n_bits: 251,
+            stop_ptr: None,
+            instances_per_component: 1,
+            signatures: HashMap::new(),
+        }
+    }
+
+    pub fn add_signature(&mut self, relocatable: Relocatable, signature: Signature) {
+        self.signatures.entry(relocatable).or_insert(signature);
+    }
+}
+
+impl SignatureBuiltinRunner {
+    pub fn initialize_segments(
+        &mut self,
+        segments: &mut MemorySegmentManager,
+        memory: &mut Memory,
+    ) {
+        self.base = segments.add(memory).segment_index
+    }
+
+    pub fn initial_stack(&self) -> Vec<MaybeRelocatable> {
+        if self.included {
+            vec![MaybeRelocatable::from((self.base, 0))]
+        } else {
+            vec![]
+        }
+    }
+
+    pub fn base(&self) -> isize {
+        self.base
+    }
+    pub fn add_validation_rule(&self, memory: &mut Memory) -> Result<(), RunnerError> {
+        let cells_per_instance = self.cells_per_instance;
+        let signatures = self.signatures.clone();
+        let rule: ValidationRule = ValidationRule(Box::new(
+            move |memory: &Memory,
+                  address: &MaybeRelocatable|
+                  -> Result<Vec<MaybeRelocatable>, MemoryError> {
+                let address = match address {
+                    MaybeRelocatable::RelocatableValue(address) => address,
+                    _ => return Err(MemoryError::AddressNotRelocatable),
+                };
+
+                let address_offset = address.offset.mod_floor(&(cells_per_instance as usize));
+                let mem_addr_sum = memory.get(&(address + 1_i32));
+                let mem_addr_less = if address.offset > 0 {
+                    memory.get(
+                        &address
+                            .sub(1)
+                            .map_err(|_| MemoryError::AddressNotRelocatable)?,
+                    )
+                } else {
+                    Ok(None)
+                };
+                let (pubkey_addr, msg_addr) = match (address_offset, mem_addr_sum, mem_addr_less) {
+                    (0, Ok(Some(_element)), _) => {
+                        let pubkey_addr = address.clone();
+                        let msg_addr = address + 1_i32;
+                        (pubkey_addr, msg_addr)
+                    }
+                    (1, _, Ok(Some(_element))) if address.offset > 0 => {
+                        let pubkey_addr = address
+                            .sub(1)
+                            .map_err(|_| MemoryError::AddressNotRelocatable)?;
+                        let msg_addr = address.clone();
+                        (pubkey_addr, msg_addr)
+                    }
+                    _ => return Ok(Vec::new()),
+                };
+
+                let (_sign, msg) = memory
+                    .get_integer(&msg_addr)
+                    .map_err(|_| MemoryError::AddressNotRelocatable)?
```

Comment:
This error management here doesn't make sense, get_integer will never return a relocatable value